### PR TITLE
Add support for user-defined functions and `cuda_compile`

### PR DESCRIPTION
**Pull request by the-slow-one**
This PR adds CMake migration rule for user-defined functions and `cuda_compile` alone with the LIT test.

~TODO~

- [x] Handle special cases in helper function (See _dpct.cmake:dpct_compile_)
- [x] Verify the LIT test for `cuda_compile`

Resolves https://jira.devtools.intel.com/browse/CTST-5674

### CODE REVIEW

File: clang/test/dpct/cmake_migation/case_018/input.cmake [hunk 1]
```diff
@@ -0,0 +1 @@
+cuda_compile(objs STATIC a.dp.cpp b.dp.cpp OPTIONS -O3 --profile)
```
Comment (the-slow-one) [inline clang/test/dpct/cmake_migation/case_018/input.cmake]:
```suggestion
cuda_compile(objs a.dp.cpp b.dp.cpp STATIC OPTIONS -O3 --profile)
```

File: clang/test/dpct/cmake_migation/case_018/expected.txt [hunk 2]
```diff
@@ -0,0 +1 @@
+dpct_compile(objs STATIC a.dp.cpp b.dp.cpp OPTIONS -O3 --profile)
```
Comment (the-slow-one) [inline clang/test/dpct/cmake_migation/case_018/expected.txt]:
```suggestion
dpct_compile(objs a.dp.cpp b.dp.cpp STATIC OPTIONS -O3 --profile)
```

File: clang/test/dpct/cmake_migation/case_016/case_016_user_defined_functions.cpp [hunk 3]
```diff
@@ -0,0 +1,10 @@
+// RUN: rm -rf %T && mkdir -p %T
+// RUN: cd %T
+// RUN: cp %S/input.cmake ./input.cmake
+// RUN: dpct -in-root ./ -out-root out  ./input.cmake --rule-file=%S/../../../../tools/dpct/DpctOptRules/cmake_script_migration_rule.yaml --migrate-cmake-script-only --cuda-include-path="%cuda-path/include"
```
Comment (tomflinda) [inline clang/test/dpct/cmake_migation/case_016/case_016_user_defined_functions.cpp]: --rule-file=%S/../../../../tools/dpct/DpctOptRules/cmake_script_migration_rule.yaml and --cuda-include-path="%cuda-path/include" could be removed.

File: clang/test/dpct/cmake_migation/case_016/expected.txt [hunk 4]
```diff
@@ -0,0 +1,13 @@
+function (user_function file1 file2)
+  message("Doing nothing in function1")
+endfunction
```
Comment (tomflinda) [inline clang/test/dpct/cmake_migation/case_016/expected.txt]: Could be removed.

File: clang/test/dpct/cmake_migation/case_016/input.cmake [hunk 5]
```diff
@@ -0,0 +1,13 @@
+function (user_function file1 file2)
+  message("Doing nothing in function1")
+endfunction
```
Comment (tomflinda) [inline clang/test/dpct/cmake_migation/case_016/input.cmake]: Could be removed.

File: clang/test/dpct/cmake_migation/case_018/case_018_cuda_compile.cpp [hunk 6]
```diff
@@ -0,0 +1,10 @@
+// RUN: rm -rf %T && mkdir -p %T
+// RUN: cd %T
+// RUN: cp %S/input.cmake ./input.cmake
+// RUN: dpct -in-root ./ -out-root out  ./input.cmake --rule-file=%S/../../../../tools/dpct/DpctOptRules/cmake_script_migration_rule.yaml --migrate-cmake-script-only --cuda-include-path="%cuda-path/include"
```
Comment (tomflinda) [inline clang/test/dpct/cmake_migation/case_018/case_018_cuda_compile.cpp]: Same as above

File: clang/tools/dpct/cmake/dpct.cmake [hunk 7]
```diff
@@ -78,3 +78,48 @@ endmacro()
 # Always set SYCL_HAS_FP16 to true to assume SYCL device to support float16
 message("dpct.cmake: SYCL_HAS_FP16 is set true by default.")
 set(SYCL_HAS_FP16 TRUE)
+
+# Replaces cuda_compile with it's equivalent SYCL repr
+macro(DPCT_COMPILE generated_files)
```
Comment (tomflinda) [inline clang/tools/dpct/cmake/dpct.cmake]: According to the definition of cuda_compile:
cuda_compile(<generated_files> <file>... [STATIC | SHARED | MODULE]
             [OPTIONS ...])
Returns a list of generated files from the input source files to be used with [add_library()](https://cmake.org/cmake/help/latest/command/add_library.html#command:add_library) or [add_executable()](https://cmake.org/cmake/help/latest/command/add_executable.html#command:add_executable).

So,for each input source file, like a.cu, b.cu, c.cu, and there should be  corresponding output files, like name_a.o, name_b.o, name_c.o (name is used for reference, not the exact name returned by cuda_compile).

While for DPCT_COMPILE, what it returns for a.dp.cpp, b.dp.cpp, c.dp.cpp?

File: clang/test/dpct/cmake_migation/case_018/input.cmake [hunk 8]
```diff
@@ -0,0 +1 @@
+cuda_compile(objs a.dp.cpp b.dp.cpp STATIC OPTIONS -O3 --profile)
```
Comment (zhiweij1) [inline clang/test/dpct/cmake_migation/case_018/input.cmake]:
```suggestion
cuda_compile(objs a.cu b.cu STATIC OPTIONS -O3 --profile)
```

File: clang/test/dpct/cmake_migation/case_016/case_016_user_defined_functions.cpp (hunk 3 above)
Comment (the-slow-one) [reply clang/test/dpct/cmake_migation/case_016/case_016_user_defined_functions.cpp]: Done

File: clang/test/dpct/cmake_migation/case_016/expected.txt (hunk 4 above)
Comment (the-slow-one) [reply clang/test/dpct/cmake_migation/case_016/expected.txt]: Done

File: clang/test/dpct/cmake_migation/case_016/input.cmake (hunk 5 above)
Comment (the-slow-one) [reply clang/test/dpct/cmake_migation/case_016/input.cmake]: Done

File: clang/test/dpct/cmake_migation/case_018/case_018_cuda_compile.cpp (hunk 6 above)
Comment (the-slow-one) [reply clang/test/dpct/cmake_migation/case_018/case_018_cuda_compile.cpp]: Done

File: clang/tools/dpct/cmake/dpct.cmake (hunk 7 above)
Comment (the-slow-one) [reply clang/tools/dpct/cmake/dpct.cmake]: `dpct_compile` use the create static or dynamic libs based on the type passed to `cuda_compile`. But I see that is not the same behaviour as `cuda_compile`. To fix it I have modified the file _dpct.cmake_ to return a list of object file path as `generated_files`. Please take a look.

File: clang/test/dpct/cmake_migration/case_018/expected.txt [hunk 9]
```diff
@@ -0,0 +1 @@
+dpct_compile(objs a.dp.cpp b.dp.cpp STATIC OPTIONS -O3 --profile)
```
Comment (tomflinda) [inline clang/test/dpct/cmake_migration/case_018/expected.txt]: "STATIC OPTIONS -O3 --profile" is specific to cuda_compile, should be removed in SYCL side.

File: clang/tools/dpct/cmake/dpct.cmake (hunk 7 above)
Comment (tomflinda) [reply clang/tools/dpct/cmake/dpct.cmake]: I'm OK to current modification.

File: clang/tools/dpct/cmake/dpct.cmake [hunk 10]
```diff
@@ -78,3 +78,24 @@ endmacro()
...
+  # list of dp.cpp source files
+  set(_sources "")
+
+  # segregate source files, library type and ignore nvcc options
```
Comment (tomflinda) [inline clang/tools/dpct/cmake/dpct.cmake]: nvcc option should be filtered in migration,  here "and ignore nvcc options" is not needed.

**Review by tomflinda — CHANGES_REQUESTED** [review]
Pls rebase the your PR to latest code repo to fix the lit test failure.

File: clang/test/dpct/cmake_migration/case_018/expected.txt (hunk 9 above)
Comment (the-slow-one) [reply clang/test/dpct/cmake_migration/case_018/expected.txt]: Done! Added more test cases.

File: clang/tools/dpct/cmake/dpct.cmake (hunk 10 above)
Comment (the-slow-one) [reply clang/tools/dpct/cmake/dpct.cmake]: Removed.

**Review by tomflinda — APPROVED** [review]
LGTM,

File: clang/tools/dpct/DpctOptRules/cmake_script_migration_rule.yaml [hunk 11]
```diff
@@ -124,6 +124,73 @@
...
+      Out: ${arg}.dp.cpp
+
+
+- Rule: rule_cuda_compile_with_no_opts
+  Kind: CMakeRule
+  Priority: Fallback
+  MatchMode: Partial
+  CmakeSyntax: rule_cuda_compile_05
+  In: cuda_compile(${libname} ${srcs})
+  Out: dpct_compile(${libname} ${srcs})
```
Comment (zhimingwang36) [inline clang/tools/dpct/DpctOptRules/cmake_script_migration_rule.yaml]: I'd prefer to dpct_helper_sycl_compile.  It gives a sense that dpct is compiler, but actually it is not.

File: clang/tools/dpct/cmake/dpct.cmake [hunk 12]
```diff
@@ -78,3 +78,15 @@ endmacro()
...
 message("dpct.cmake: SYCL_HAS_FP16 is set true by default.")
 set(SYCL_HAS_FP16 TRUE)
+
+# Replaces cuda_compile with it's equivalent SYCL repr
+macro(DPCT_COMPILE generated_files)
+  DPCT_GET_SOURCES(_sources ${ARGN})
+
+  # can't continue without list of source files
+  if("${_sources}" STREQUAL "")
+    message(FATAL "dpct: failed to find source files while migrating `cuda_compile`")
```
Comment (zhimingwang36) [inline clang/tools/dpct/cmake/dpct.cmake]: It's build time that the message in Line88 will be print out (not during migration),  so how about update the message to "Failed to find source files when run DPCT_HELPER_SYCL_COMPILE macro".

File: clang/tools/dpct/cmake/dpct.cmake [hunk 13]
```diff
@@ -78,3 +78,15 @@ endmacro()
 # Always set SYCL_HAS_FP16 to true to assume SYCL device to support float16
 message("dpct.cmake: SYCL_HAS_FP16 is set true by default.")
 set(SYCL_HAS_FP16 TRUE)
+
+# Replaces cuda_compile with it's equivalent SYCL repr
```
Comment (zhimingwang36) [inline clang/tools/dpct/cmake/dpct.cmake]: L82 comment is not correct, it should show the fun of DPCT_HELPER_SYCL_COMPILE

File: clang/tools/dpct/cmake/dpct.cmake [hunk 14]
```diff
@@ -78,3 +78,15 @@ endmacro()
...
+macro(DPCT_COMPILE generated_files)
```
Comment (zhimingwang36) [inline clang/tools/dpct/cmake/dpct.cmake]:
```suggestion
macro(DPCT_HELPER_SYCL_COMPILE generated_files)
```

**Review by zhimingwang36 — CHANGES_REQUESTED** [review]
LGTM

File: clang/tools/dpct/DpctOptRules/cmake_script_migration_rule.yaml (hunk 11 above)
Comment (the-slow-one) [reply clang/tools/dpct/DpctOptRules/cmake_script_migration_rule.yaml]: Sounds correct! Modified.

File: clang/tools/dpct/cmake/dpct.cmake (hunk 12 above)
Comment (the-slow-one) [reply clang/tools/dpct/cmake/dpct.cmake]: Update the message to the following:
> Failed to find the source files while running the macro 'DPCT_HELPER_SYCL_COMPILE'

File: clang/tools/dpct/cmake/dpct.cmake (hunk 13 above)
Comment (the-slow-one) [reply clang/tools/dpct/cmake/dpct.cmake]: That's correct! Updated the comment to the following

> Return the list of object file paths generated for the given SYCL source files

File: clang/tools/dpct/cmake/dpct.cmake (hunk 14 above)
Comment (the-slow-one) [reply clang/tools/dpct/cmake/dpct.cmake]: Done

**Review by zhimingwang36 — APPROVED** [review]
LGTM

**Pull request merged**